sql: remove dead interleave code

postamar opened the pull request:
sql: remove dead interleave code
    
    In the previous release, we disabled interleaved tables and indexes.
    Therefore, at this point in time, all code pertaining to this feature is
    effectively dead. This commit removes a lot of it by removing the
    catalog methods and their callers.
    
    The remaining dead code is now located in the SQL grammar definition and
    the plan nodes, and is removed in the subsequent commit.
    
    Release note: None


    sql: remove INTERLEAVE IN PARENT from grammar
    
    This commit follows up on the previous one by removing INTERLEAVE
    clauses from the SQL grammar itself.
    
    Release note (sql change): Removed INTERLEAVE IN PARENT.

Commit:
sql: remove dead interleave code

In the previous release, we disabled interleaved tables and indexes.
Therefore, at this point in time, all code pertaining to this feature is
effectively dead. This commit removes a lot of it by removing the
catalog methods and their callers.

The remaining dead code is now located in the SQL grammar definition and
the plan nodes, and is removed in the subsequent commit.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/70618)
<!-- Reviewable:end -->

Commit:
sql: remove INTERLEAVE IN PARENT from grammar

This commit follows up on the previous one by removing INTERLEAVE
clauses from the SQL grammar itself.

Release note (sql change): Removed INTERLEAVE IN PARENT.

postamar commented:
This is not ready to merge, but it's ready to review, if anyone dares. One aspect of my change I'm not too sure of is if I was supposed to do something in particular to deprecate the cluster settings instead of outright removing those which I removed.

I appreciate that this diff is hard to review. I've broken it down into two commits but the substance of the change is in the first one. There's a lot of noise in there, but there are also test cases which I outright removed because I felt they no longer made sense, that kind of thing. What I'm trying to say is that this should probably be reviewed quite carefully despite the noise. I tried to break the change down further but everything is quite intertwined.

yuzefovich commented:
I have #70239 that removes the interleaved logic from the fetchers - it'd be probably good to rebase this PR on top of that one since #70239 has already been approved (I was thinking of waiting till 21.2.0 is out, just in case we need to backport something).

postamar commented:
Excellent! I'll do just that, once the release is done. I may even hold off until your PR is merged, depending on circumstances.

yuzefovich reviewed (commented):
Thanks for working on this! It's exciting to see all this code go away.

> Reviewed 110 of 111 files at r1, 48 of 48 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-:-Mn2P46tDYHmJ1J80Ll7:bq9pm5w)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @postamar, and @tbg)

___
*[docs/generated/sql/bnf/BUILD.bazel, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2Ngio3KKgPyuXgiKr:-Mn2Ngio3KKgPyuXgiKs:b-t7akg9) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4a177d695c3a984a73766b54bce616dfcb0f4fc/docs/generated/sql/bnf/BUILD.bazel#L1)):*
> ```Starlark
> # There should be an entry in this list for every .bnf in this directory (omit
> ```

Seems like this changed more than necessary, what was the reason?
___
*[pkg/ccl/backupccl/backupresolver/targets.go, line 635 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2Myg90dv6BEdaT6HH:-Mn2Myg90dv6BEdaT6HI:b-vsufmy) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/ccl/backupccl/backupresolver/targets.go#L635)):*
> ```Go
> 	}
> 
> 	sort.Slice(matched.Descs, func(i, j int) bool { return matched.Descs[i].GetID() < matched.Descs[j].GetID() })
> ```

Do we still need to sort here? The old comment implied the sort was only needed because of the interleaved tables.
___
*[pkg/clusterversion/cockroach_versions.go, line 394 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2NFHLC1Pg7q6KI01N:-Mn2NFHLC1Pg7q6KI01O:bxyb7p1) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4a177d695c3a984a73766b54bce616dfcb0f4fc/pkg/clusterversion/cockroach_versions.go#L394)):*
> ```Go
> 	},
> 	{
> 		Key:     PreventNewInterleavedTables,
> ```

I don't know anything about removing versions, so I'll defer to someone else to look at this. Maybe @irfansharif or @ajwerner. 
___
*[pkg/sql/backfill.go, line 859 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn1RNAf98SY8mnc4lm1:-Mn1RNAf98SY8mnc4lm2:b-ekac3b) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4a177d695c3a984a73766b54bce616dfcb0f4fc/pkg/sql/backfill.go#L859)):*
> ```Go
> 	}
> 	for _, idx := range dropped {
> 		var resume roachpb.Span
> ```

IIUC correctly `resume` can be removed too - this was introduced only to support dropping of the interleaved indexes, and it'll be always unset. Additionally, the `for` loop also now seems unnecessary since we always will do just a single iteration.
___
*[pkg/sql/crdb_internal.go, line 2419 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn1WuiiA5pzi9xunNRy:-Mn1WuiiA5pzi9xunNRz:b-u2n4x1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/crdb_internal.go#L2419)):*
> ```Go
> 	})
> 
> func showAlterStatement(
> ```

nit: I think this method is called in a single place, should we inline it then?
___
*[pkg/sql/show_create.go, line 62 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn1Z5Oc7JTQlnkB3m4l:-Mn1Z5Oc7JTQlnkB3m4m:b-cr0gld) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/show_create.go#L62)):*
> ```Go
> // TABLE statement used to create the given table.
> //
> // The names of the tables references by foreign keys are prefixed by their own
> ```

nit: `s/references/referenced/`.
___
*[pkg/sql/tablewriter_delete.go, line 67 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn1ZiqO5K9GKKlEVh9w:-Mn1ZiqO5K9GKKlEVh9x:b-talig1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/tablewriter_delete.go#L67)):*
> ```Go
> // zero resume-span. After a chunk is deleted a new resume-span is returned.
> //
> // limit is a limit on either the number of keys deleted in the operation.
> ```

nit: `s/on either the number/on the number/`.
___
*[pkg/sql/catalog/descriptor.go, line 488 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn1_XTp8YohhwuGm_GU:-Mn1_XTp8YohhwuGm_GV:b458dmh) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4a177d695c3a984a73766b54bce616dfcb0f4fc/pkg/sql/catalog/descriptor.go#L488)):*
> ```Go
> 	GetNextFamilyID() descpb.FamilyID
> 
> 	// HasColumnBackfillMutation returns whether the table has any queued column
> ```

nit: why remove the comment?
___
*[pkg/sql/catalog/descpb/structured.proto, line 516 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2CWA01nF5JPuioeO1:-Mn2CWA01nF5JPuioeO2:byriwpt) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/catalog/descpb/structured.proto#L516)):*
> ```protobuf
>   // Interleave, if it's not the zero value, describes how this index's data is
>   // interleaved into another index's data.
>   optional InterleaveDescriptor interleave = 11 [(gogoproto.nullable) = false, deprecated = true];
> ```

Can we just delete these two fields and mark the corresponding ordinals as `reserved`? I'm guessing not because of the backups.
___
*[pkg/sql/logictest/testdata/logic_test/pg_builtins, line 541 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2EvVeBGN4sqsQ1hta:-Mn2EvVeBGN4sqsQ1htb:bfxcim3) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4a177d695c3a984a73766b54bce616dfcb0f4fc/pkg/sql/logictest/testdata/logic_test/pg_builtins#L541)):*
> ```
> indexed_b_d_idx  {2,4}  4  2
> 
> # information_schema._pg_numeric_precision and information_schema._pg_numeric_precision_radix
> ```

Why is this chunk removed?
___
*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 3436 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2F3Od5GFhaJEbiUta:-Mn2F3Od5GFhaJEbiUtb:bfxcim3) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4a177d695c3a984a73766b54bce616dfcb0f4fc/pkg/sql/logictest/testdata/logic_test/pg_catalog#L3436)):*
> ```
> oid     typname        typnamespace  typowner  typlen  typbyval  typtype
> 
> query OTOOIBT colnames
> ```

Why is this chunk removed?
___
*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 5420 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2F9XVBygyC9T2Lhp8:-Mn2F9XVBygyC9T2Lhp9:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4a177d695c3a984a73766b54bce616dfcb0f4fc/pkg/sql/logictest/testdata/logic_test/pg_catalog#L5420)):*
> ```
> regression_70180  NULL
> 
> # Testing pg_statistic_ext
> ```

ditto
___
*[pkg/sql/opt/exec/execbuilder/mutation.go, line 603 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2FGK15lbPbdqRfTl0:-Mn2FGK15lbPbdqRfTl1:b-pnl1rj) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/opt/exec/execbuilder/mutation.go#L603)):*
> ```Go
> 		if maxRows, ok := b.indexConstraintMaxResults(&scan.ScanPrivate, scan.Relational()); ok {
> 			if maxKeys := maxRows * uint64(tab.FamilyCount()); maxKeys <= row.TableTruncateChunkSize {
> ```

nit: an extra new line.
___
*[pkg/sql/opt/testutils/opttester/testfixtures/tpcc_schema, line 62 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2ORDo5AJDRevzqsMT:-Mn2ORDo5AJDRevzqsMU:bo7uf7g) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/opt/testutils/opttester/testfixtures/tpcc_schema#L62)):*
> ```
>     index customer_idx (c_w_id, c_d_id, c_last, c_first),
>     foreign key (c_w_id, c_d_id) references district (d_w_id, d_id)
> ) interleave in parent district (c_w_id, c_d_id)
> ```

super nit: seems like these couple of changes belong better in the first commit.
___
*[pkg/sql/opt/xform/testdata/rules/join, line 255 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2FY_CAazQJhiPL4xQ:-Mn2FY_CAazQJhiPL4xR:b-puqpo4) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4a177d695c3a984a73766b54bce616dfcb0f4fc/pkg/sql/opt/xform/testdata/rules/join#L255)):*
> ```
>   PRIMARY KEY(pid1, cid1)
> )
> INTERLEAVE IN PARENT parent1 (pid1)
> ```

I wonder whether these tests should be converted to use foreign keys instead of interleaving?
___
*[pkg/sql/randgen/schema.go, line 80 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2Fs0n9jxa1iE6Nu1l:-Mn2Fs0o4R6gIfG-ER10:bglpx4a) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/randgen/schema.go#L80)):*
> ```Go
> 	tables := make([]tree.Statement, num)
> 	for i := 0; i < num; i++ {
> 		t := RandCreateTableWithColumnIndexNumberGenerator(rng, prefix, i+1, nil /* generateColumnIndexNumber */)
> ```

nit: should we use `RandCreateTable` here?
___
*[pkg/sql/row/fetcher.go, line 212 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2Jy4h7P5GT9WrxrvI:-Mn2Jy4h7P5GT9WrxrvJ:b-2usu5x) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/row/fetcher.go#L212)):*
> ```Go
> 
> 	// True if the index key must be decoded.
> 	// If there is more than one table, the index key must always be decoded.
> ```

nit: no need to change this comment since `row.Fetcher` now supports only a single table.
___
*[pkg/sql/rowenc/index_encoding.go, line 506 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2KgOW-E3KWBeSta3G:-Mn2KgOW-E3KWBeSta3H:b-8c17o9) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/rowenc/index_encoding.go#L506)):*
> ```Go
> 	// bookkeeping. We could improve it to n with a prefix tree of components.
> 
> 	interleaves := append(make([]catalog.Index, 0, len(desc.ActiveIndexes())), desc.ActiveIndexes()...)
> ```

I think the logic here could be simplified further (maybe the `for` loop below can be removed entirely?).
___
*[pkg/sql/rowenc/index_encoding.go, line 606 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2L6gGBYyerBoMriXM:-Mn2L6gGBYyerBoMriXN:b9hh32k) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/rowenc/index_encoding.go#L606)):*
> ```Go
> 	// interleavedSentinel is actually next, then this key is for a child
> 	// table.
> 	if _, ok := encoding.DecodeIfInterleavedSentinel(key); ok {
> ```

I think this can be removed. As a result we could remove `DecodeIndexKeyWithoutTableIDIndexIDPrefix` entirely and just inline `DecodeKeyVals` instead.
___
*[pkg/sql/rowenc/index_encoding.go, line 1527 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2LQYy6EaCjdXSCxnA:-Mn2LQYy6EaCjdXSCxnB:bk48tj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/rowenc/index_encoding.go#L1527)):*
> ```Go
> 	desc catalog.TableDescriptor, index catalog.Index,
> ) (signatures [][]byte, err error) {
> 	// signatures contains the slice reference to the signature of every
> ```

nit: the comment needs an update.
___
*[pkg/sql/rowenc/index_encoding.go, line 1533 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2LUVTEZjRbuxVnWuY:-Mn2LUVTEZjRbuxVnWuZ:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/rowenc/index_encoding.go#L1533)):*
> ```Go
> 	// fullSignature is the backing byte slice for each individual signature
> 	// as it buffers each block of table and index IDs.
> 	// We eagerly allocate 4 bytes for each of the two IDs per ancestor
> ```

ditto
___
*[pkg/sql/rowenc/index_encoding.go, line 1554 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2LZOz0bkS4ZLPBqQw:-Mn2LZOz0bkS4ZLPBqQx:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/rowenc/index_encoding.go#L1554)):*
> ```Go
> // if uncertain, pass in true to 'overestimate' the maxKeyTokens.
> //
> // In general, a key belonging to an interleaved index grandchild is encoded as:
> ```

ditto
___
*[pkg/sql/span/span_builder.go, line 87 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2LznA6LrzTS_Z9hRZ:-Mn2LznA6LrzTS_Z9hR_:b-kbqv2o) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/span/span_builder.go#L87)):*
> ```Go
> 
> 	// Set up the interstices for encoding interleaved tables later.
> 	s.interstices[0] = s.KeyPrefix
> ```

I think we can simplify these `interstices` further. Feel free to leave `TODO(yuzefovich)` about it.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70618)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @irfansharif, and @postamar)

___
*[pkg/clusterversion/cockroach_versions.go, line 394 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2NFHLC1Pg7q6KI01N:-Mn2SAJY0XFLe5yvFO0t:b-ij8nm2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4a177d695c3a984a73766b54bce616dfcb0f4fc/pkg/clusterversion/cockroach_versions.go#L394)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I don't know anything about removing versions, so I'll defer to someone else to look at this. Maybe @irfansharif or @ajwerner. 
</blockquote></details>

In serverless, tenant pods can trail the KV host cluster by one release. So when KV runs 21.2, tenants can still run 21.1.  I am wagering a guess that the migrations attached to these versions are entirely on the tenant side, and so things should be fine (a 21.1 tenant can have interleaved tables, but the 21.2 host cluster couldn't care less).

As for removing settings, current `master` is going to be the 22.1 release and so will only ever be run in cluster versions 21.2 and up. The version here is pre-21.2, so this should be good to remove.


<!-- Sent from Reviewable.io -->

postamar reviewed (commented):
Thanks for taking time to review! I bet the excitement of seeing this go away is largely tempered by the drudgery of plowing through 10k line diff.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-:-Mn7Sh7T6zeQwzHHmkTi:beg5436)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @irfansharif, and @yuzefovich)

___
*[docs/generated/sql/bnf/BUILD.bazel, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2Ngio3KKgPyuXgiKr:-Mn7T-z51F-7CSyCfaUa:b-w37a5w) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4a177d695c3a984a73766b54bce616dfcb0f4fc/docs/generated/sql/bnf/BUILD.bazel#L1)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Seems like this changed more than necessary, what was the reason?
</blockquote></details>

The reason was `make bazel-generate` followed by a careless `git add`. Fixed.
___
*[pkg/ccl/backupccl/backupresolver/targets.go, line 635 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2Myg90dv6BEdaT6HH:-Mn7T9833ebSUggou-f_:budu0fg) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/ccl/backupccl/backupresolver/targets.go#L635)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Do we still need to sort here? The old comment implied the sort was only needed because of the interleaved tables.
</blockquote></details>

Agreed. I had a look at the call sites just to be sure: I'm reasonably confident that it's of no use now. Removed.
___
*[pkg/clusterversion/cockroach_versions.go, line 394 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2NFHLC1Pg7q6KI01N:-Mn7URYE9HBSI8vpHcoZ:bb8hvjl) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4a177d695c3a984a73766b54bce616dfcb0f4fc/pkg/clusterversion/cockroach_versions.go#L394)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

In serverless, tenant pods can trail the KV host cluster by one release. So when KV runs 21.2, tenants can still run 21.1.  I am wagering a guess that the migrations attached to these versions are entirely on the tenant side, and so things should be fine (a 21.1 tenant can have interleaved tables, but the 21.2 host cluster couldn't care less).

As for removing settings, current `master` is going to be the 22.1 release and so will only ever be run in cluster versions 21.2 and up. The version here is pre-21.2, so this should be good to remove.
</blockquote></details>

Correct, these migrations are tenant migrations. This seems to be the case for all SQL-related migrations, from what I can tell.
___
*[pkg/sql/backfill.go, line 859 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn1RNAf98SY8mnc4lm1:-Mn7UlJ5EhbycAu1uyRm:bhmhk5s) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4a177d695c3a984a73766b54bce616dfcb0f4fc/pkg/sql/backfill.go#L859)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

IIUC correctly `resume` can be removed too - this was introduced only to support dropping of the interleaved indexes, and it'll be always unset. Additionally, the `for` loop also now seems unnecessary since we always will do just a single iteration.
</blockquote></details>

Nice catch! Done.
___
*[pkg/sql/crdb_internal.go, line 2419 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn1WuiiA5pzi9xunNRy:-Mn7Unra16fMu3WSE8Zw:bhyw2uk) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/crdb_internal.go#L2419)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: I think this method is called in a single place, should we inline it then?
</blockquote></details>

The call site is already quite deeply nested, I'll leave it as it is, if you don't mind.
___
*[pkg/sql/show_create.go, line 62 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn1Z5Oc7JTQlnkB3m4l:-Mn7UtbbDvB9-6nJpLRF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/show_create.go#L62)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: `s/references/referenced/`.
</blockquote></details>

Done.
___
*[pkg/sql/tablewriter_delete.go, line 67 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn1ZiqO5K9GKKlEVh9w:-Mn7Uv0S3KC3AhbYUcDU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/tablewriter_delete.go#L67)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: `s/on either the number/on the number/`.
</blockquote></details>

Done.
___
*[pkg/sql/catalog/descriptor.go, line 488 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn1_XTp8YohhwuGm_GU:-Mn7UwYF1vh7s_r1pxOH:bt4ehj1) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4a177d695c3a984a73766b54bce616dfcb0f4fc/pkg/sql/catalog/descriptor.go#L488)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: why remove the comment?
</blockquote></details>

No good reason, this was a mistake. Good catch.
___
*[pkg/sql/catalog/descpb/structured.proto, line 516 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2CWA01nF5JPuioeO1:-Mn7V2it0D2fnnTos1FN:bblvw7f) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/catalog/descpb/structured.proto#L516)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Can we just delete these two fields and mark the corresponding ordinals as `reserved`? I'm guessing not because of the backups.
</blockquote></details>

Yes, it's because of the backups. More generally, we need to check dead fields such as this one during descriptor validation to make sure that it's empty. Consequently nothing in `structured.proto` can ever be deleted. It's not such a big deal as it sounds, this file doesn't change that much these days.
___
*[pkg/sql/logictest/testdata/logic_test/pg_builtins, line 541 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2EvVeBGN4sqsQ1hta:-Mn7W0pGAMEaLvlPpPT-:b-p4q1gp) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4a177d695c3a984a73766b54bce616dfcb0f4fc/pkg/sql/logictest/testdata/logic_test/pg_builtins#L541)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Why is this chunk removed?
</blockquote></details>

No good reason. Good catch.
___
*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 3436 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2F3Od5GFhaJEbiUta:-Mn7W4uSAAPdez98ZD-j:b-p4q1gp) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4a177d695c3a984a73766b54bce616dfcb0f4fc/pkg/sql/logictest/testdata/logic_test/pg_catalog#L3436)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Why is this chunk removed?
</blockquote></details>

No good reason. Good catch.
___
*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 5420 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2F9XVBygyC9T2Lhp8:-Mn7W6fg3eo4ulfq0TX0:b-ucfkn4) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4a177d695c3a984a73766b54bce616dfcb0f4fc/pkg/sql/logictest/testdata/logic_test/pg_catalog#L5420)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Thanks.
___
*[pkg/sql/opt/testutils/opttester/testfixtures/tpcc_schema, line 62 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2ORDo5AJDRevzqsMT:-Mn7W8AJBC2flF3R6DRp:bbwqedd) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/opt/testutils/opttester/testfixtures/tpcc_schema#L62)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

super nit: seems like these couple of changes belong better in the first commit.
</blockquote></details>

I'm not sure I agree, but don't feel too strongly about this. I'll leave it in the current commit at least out of laziness, unless someone has a strong opinion.
___
*[pkg/sql/opt/xform/testdata/rules/join, line 255 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2FY_CAazQJhiPL4xQ:-Mn7WV0C0hmYjoX4gUJ1:b7y0ob5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4a177d695c3a984a73766b54bce616dfcb0f4fc/pkg/sql/opt/xform/testdata/rules/join#L255)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I wonder whether these tests should be converted to use foreign keys instead of interleaving?
</blockquote></details>

I'm surprised they weren't broken in the first place, DDLs with INTERLEAVE should trigger errors nowadays.
___
*[pkg/sql/randgen/schema.go, line 80 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2Fs0n9jxa1iE6Nu1l:-Mn7WsGa6eph5uCnKsnk:b-xvohb6) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/randgen/schema.go#L80)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: should we use `RandCreateTable` here?
</blockquote></details>

You're right. Done.
___
*[pkg/sql/row/fetcher.go, line 212 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2Jy4h7P5GT9WrxrvI:-Mn7Wv6q1vppFG91wgzE:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/row/fetcher.go#L212)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: no need to change this comment since `row.Fetcher` now supports only a single table.
</blockquote></details>

Fixed.
___
*[pkg/sql/rowenc/index_encoding.go, line 506 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2KgOW-E3KWBeSta3G:-Mn7Wx4o3LPiTEO7hawb:bx1brgf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/rowenc/index_encoding.go#L506)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I think the logic here could be simplified further (maybe the `for` loop below can be removed entirely?).
</blockquote></details>

I think so too. I don't know why I didn't do it. Perhaps I was waiting for your changes? In any case, it's done. 
___
*[pkg/sql/rowenc/index_encoding.go, line 606 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2L6gGBYyerBoMriXM:-Mn7X56z5mWCnuXZR15o:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/rowenc/index_encoding.go#L606)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I think this can be removed. As a result we could remove `DecodeIndexKeyWithoutTableIDIndexIDPrefix` entirely and just inline `DecodeKeyVals` instead.
</blockquote></details>

Done.
___
*[pkg/sql/rowenc/index_encoding.go, line 1527 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2LQYy6EaCjdXSCxnA:-Mn7X6lN9AXGlYbnV4YZ:b-mgsme) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/rowenc/index_encoding.go#L1527)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: the comment needs an update.
</blockquote></details>

As a matter of fact, this function isn't used anywhere except in tests, thanks to your recent changes. I removed the lot.
___
*[pkg/sql/rowenc/index_encoding.go, line 1533 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2LUVTEZjRbuxVnWuY:-Mn7XDuy70vhSu43R1r9:b-q64pkt) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/rowenc/index_encoding.go#L1533)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

See above.
___
*[pkg/sql/rowenc/index_encoding.go, line 1554 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2LZOz0bkS4ZLPBqQw:-Mn7XONw13rpE6wL0tD3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/rowenc/index_encoding.go#L1554)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.
___
*[pkg/sql/span/span_builder.go, line 87 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2LznA6LrzTS_Z9hRZ:-Mn7YCQGC007zLZnUPSt:b-vy3wk8) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/span/span_builder.go#L87)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I think we can simplify these `interstices` further. Feel free to leave `TODO(yuzefovich)` about it.
</blockquote></details>

Done. I don't know what these are, so I don't dare touch them.
___
*[pkg/sql/opt/exec/execbuilder/mutation.go, line 603 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2FGK15lbPbdqRfTl0:-Mn7YIWT-1u5V7tnBan-:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/opt/exec/execbuilder/mutation.go#L603)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: an extra new line.
</blockquote></details>

Fixed.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
LOL

I think last thing is that we need to add the removed cluster setting into `retiredSettings` map. Otherwise, <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 111 files at r1, 76 of 76 files at r3, 49 of 49 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-:-Mn8CB881ONZAwjDp58p:b-qmy19l)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @irfansharif, and @postamar)

___
*[pkg/sql/opt/xform/testdata/rules/join, line 255 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2FY_CAazQJhiPL4xQ:-Mn8C-wBDNHxNPZG3sTQ:buuyfd2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4a177d695c3a984a73766b54bce616dfcb0f4fc/pkg/sql/opt/xform/testdata/rules/join#L255)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I'm surprised they weren't broken in the first place, DDLs with INTERLEAVE should trigger errors nowadays.
</blockquote></details>

This test file probably runs through a special optimizer setup, without really executing the queries, so we weren't hitting those errors.
___
*[pkg/sql/rowenc/index_encoding.go, line 1527 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2LQYy6EaCjdXSCxnA:-Mn8AuT0BXJ9mZcmIFYu:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8648864ac88150ec341595f90e55283159c67ae/pkg/sql/rowenc/index_encoding.go#L1527)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

As a matter of fact, this function isn't used anywhere except in tests, thanks to your recent changes. I removed the lot.
</blockquote></details>

:thumbsup: 
___
*[pkg/sql/rowenc/index_encoding.go, line 509 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn8AZK8-Vw5njXRQhp9:-Mn8AZK9EL8hPTSV9QPY:b-1twb81) ([raw file](https://github.com/cockroachdb/cockroach/blob/6baaaf553a6608c4df4c6bc606239efffcb86a50/pkg/sql/rowenc/index_encoding.go#L509)):*
> ```Go
> 	if tableID != desc.GetID() {
> 		return 0, nil, errors.Errorf(
> 			"Unexpected table ID %d, expected %d instead.", tableID, desc.GetID())
> ```

super nit: I think we use lower-case first word in the main error message.


<!-- Sent from Reviewable.io -->

postamar reviewed (commented):
Oh, I forgot all about `retiredSettings`, my bad. Done.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-:-MnBK3Sm2yDuvdRNx83j:b-bf896h)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @ajwerner, @irfansharif, and @yuzefovich)

___
*[pkg/sql/opt/xform/testdata/rules/join, line 255 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn2FY_CAazQJhiPL4xQ:-MnBKPIa8a3qZHFJwPDF:bxxx4gf) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4a177d695c3a984a73766b54bce616dfcb0f4fc/pkg/sql/opt/xform/testdata/rules/join#L255)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

This test file probably runs through a special optimizer setup, without really executing the queries, so we weren't hitting those errors.
</blockquote></details>

That makes sense.
___
*[pkg/sql/rowenc/index_encoding.go, line 509 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/70618#-Mn8AZK8-Vw5njXRQhp9:-MnBKRg3-J1EHocAZa9_:b-bcl1c7) ([raw file](https://github.com/cockroachdb/cockroach/blob/6baaaf553a6608c4df4c6bc606239efffcb86a50/pkg/sql/rowenc/index_encoding.go#L509)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

super nit: I think we use lower-case first word in the main error message.
</blockquote></details>

I'm never sure about this. Done.

In fact in that case I should remove the period also. I'll push that change at the next opportunity.


<!-- Sent from Reviewable.io -->

postamar commented:
bors r+

postamar commented:
This is so cathartic. Thanks @yuzefovich !

yuzefovich commented:
I wonder whether we should close #69150 given that this has been merged?

postamar commented:
You're right, we should. At the time that I started working on this PR, I didn't intend to remove all remaining interleave code, only that which involved `NumAncestors()` and other interleave-related catalog descriptor methods, so I didn't issue-link. I since forgot to update.
